colcon_ros_domain_id_coordinator/shell/ros_domain_id.py

```diff
@@ -0,0 +1,77 @@
+# Copyright 2023 Open Source Robotics Foundation, Inc.
+# Licensed under the Apache License, Version 2.0
+
+from asyncio import Queue
+from asyncio import QueueEmpty
+import os
+from pathlib import Path
+from random import shuffle
+
+from colcon_core.plugin_system import satisfies_version
+from colcon_core.plugin_system import SkipExtensionException
+from colcon_core.shell import logger
+from colcon_core.shell import ShellExtensionPoint
+
+
+class ROSDomainIDShell(ShellExtensionPoint):
+    """Set a different ROS_DOMAIN_ID environment variable for each task."""
+
+    # The priority should be higher than any usable shells
+    PRIORITY = 900
+
+    def __init__(self):  # noqa: D107
+        super().__init__()
+        satisfies_version(ShellExtensionPoint.EXTENSION_POINT_VERSION, '^2.2')
+
+        self.free_ids = Queue()
+
+        all_ids = []
+        all_ids.extend(range(1, 102))
+        all_ids.extend(range(215, 233))
+        shuffle(all_ids)
+        for i in all_ids:
+            self.free_ids.put_nowait(str(i))
+
+    def get_file_extensions(self):  # noqa: D102
+        return ()
+
+    def create_prefix_script(self, prefix_path, merge_install):  # noqa: D102
+        return []
+
+    def create_package_script(  # noqa: D102
+        self, prefix_path, pkg_name, hooks
+    ):
+        return []
+
+    def create_hook_set_value(  # noqa: D102
+        self, env_hook_name, prefix_path, pkg_name, name, value,
+    ):
+        return Path('ros_domain_id')
+
+    def create_hook_append_value(  # noqa: D102
+        self, env_hook_name, prefix_path, pkg_name, name, subdirectory,
+    ):
+        return Path('ros_domain_id')
+
+    def create_hook_prepend_value(  # noqa: D102
+        self, env_hook_name, prefix_path, pkg_name, name, subdirectory,
+    ):
+        return Path('ros_domain_id')
+
+    async def generate_command_environment(  # noqa: D102
+        self, task_name, build_base, dependencies,
+    ):
+        try:
+            task_id = self.free_ids.get_nowait()
+        except QueueEmpty:
+            logger.warn(f"No free ROS_DOMAIN_ID to assign for '{task_name}'")
+            os.environ.pop('ROS_DOMAIN_ID', None)
+            task_id = None
+        else:
+            os.environ['ROS_DOMAIN_ID'] = task_id
+            logger.debug(
+                f"Allocated ROS_DOMAIN_ID={task_id} for '{task_name}'")
+            self.free_ids.put_nowait(task_id)
```

Comment:
Correct, that's what I was trying to say. We can't do this perfectly right now, but this simple approach would require a single task taking the same time as over a hundred other tasks combined before there could possibly be a collision, so it's pretty robust as-is.

I have some ideas on how we can do this more robustly, but it will require changes to colcon-core.

Reply:
Yeah, after thinking about it for a while, I agree with you; this will work fairly well for most common machines, and certainly in CI.  So I'm totally fine with this as a limitation for now, though I do think it is worth documenting that limitation (and the reason for it) in the README or similar.

Reply:
I tried to capture the limitation in the README.rst. Please feel free to suggest changes to the wording.